File: src/arrays/groupedanagram/grouped_anagram.go
package main

import (
	"container/list"
	"fmt"
)

func groupAnagram(strs []string) [][]string {

	// We create a frequency map that will hold the
	// Key: .toString() of the frequencies -> We do toString because it's easier to create a hash
	// Value: Matching string
	// we are using a list because the length would grow dynamically
	frequencyMap := make(map[string]list.List)

	// Loop through all the strings available O(m)
	for _, value := range strs {

		// Create an array that will be used to hold the frequencies
		var store [26]int

		// Loop through all the string chars and extract each letter frequency
		// O(n): Total = O(m * n)
		for i := 0; i < len(value); i++ {
			store[value[i]-byte('a')]++
		}

		// Convert the array into a string
		storeString := fmt.Sprint(store)

		// Get the current value and  confirm if it already existed
		// IIf exists then another anagram is present
		existingValue, exists := frequencyMap[storeString]

		if !exists {
			var valueList = list.New()
			valueList.PushFront(value)
			frequencyMap[storeString] = *valueList
		} else {
			// If key exists  the append the current string value
			existingValue.PushFront(value)
			// Update the key value with the updated list
			frequencyMap[storeString] = existingValue
		}
	}

	// Create a multidimensional array to hold the values
	valuesArray := make([][]string, len(frequencyMap))

	// Tracks the row index
	index := 0
	for _, values := range frequencyMap {
		// For each row, we create a colum with the size of grouped values
		valuesArray[index] = make([]string, values.Len())

		// Tracks the colum index
		columIndex := 0
		for val := values.Front(); val != nil; val = val.Next() {
			// Populate the columns
			valuesArray[index][columIndex] = val.Value.(string)

			// update the column index
			columIndex++
		}

		// Update the row index
		index++
	}

	// return the values
	return valuesArray
}

func main() {
	// Expect [["hat"],["act", "cat"],["stop", "pots", "tops"]]
	fmt.Println(groupAnagram([]string{"act", "pots", "tops", "cat", "stop", "hat"}))

	// Expect [["x"]]
	fmt.Println(groupAnagram([]string{"x"}))

	// Expect [[""]]
	fmt.Println(groupAnagram([]string{""}))
}

File: src/arrays/groupedanagram/grouped_anagram.rs
use std::collections::HashMap;

fn group_anagram(strs: &[&str]) -> Vec<Vec<String>> {

    // We create a frequency map that will hold the
    // Key: .toString() of the frequencies -> We do toString because it's easier to create a hash
    // Value: Matching string
    let mut frequency_map = HashMap::new();

    for str in strs.iter() {

        // We create an array with zeros for all the 26 alphabetic characters
        let mut store: [u8 ; 26] = [0; 26];

        for c in str.chars() {
            // Get the UTF-8 index if the character
            let index :  u8 = c as u8 - 'a' as u8;
            // Get the char count, initially should be zero
            let current_value: u8 = store[index as usize];
            // Add to the char count
            store[index as usize] = current_value + 1;
        }

        // Create a to_string version of the array to be used as the key
        // If two strings are anagram they should have equal store values
        let store_to_string = format!("{:?}",store);

        // Check if there's an existing key
        match frequency_map.get(&store_to_string) {
            // If not
            None => {
                // Create a vector to store the new str
                let value_list : Vec<String> = vec![String::from(*str)];
                // Add it to the frequency map
                frequency_map.insert(store_to_string, value_list);
            },
            // If exists
            Some(value_list) => {
                // Clone the array and make it mutable so that we can add the newly matched string
                let  mut mut_value_list : Vec<String> = value_list.clone();
                // Add the matched string
                mut_value_list.push(String::from(*str));
                // Update the array
                frequency_map.insert(store_to_string, mut_value_list);
            }
        }
    }

    // Create the result object, array of arrays
    let mut result : Vec<Vec<String>> = Vec::new();

    // Fill the result with the matched strings
    for (_, value) in  frequency_map {
        result.push(value);
    }

    // Return result
    result
}

fn main(){

    // Expect [["hat"],["act", "cat"],["stop", "pots", "tops"]]
    println!("{:?}",group_anagram(&["act", "pots", "tops", "cat", "stop", "hat"]));

    // Expect [["x"]]
    println!("{:?}",group_anagram(&["x"]));

    // Expect [[""]]
    println!("{:?}",group_anagram(&[""]));
}
